crates/turbopack-ecmascript/src/references/async_module.rs

```diff
@@ -124,19 +113,23 @@ impl AsyncModuleScc {
     async fn is_async(self: Vc<Self>) -> Result<Vc<bool>> {
         let this = self.await?;
 
-        let mut bools = Vec::new();
-
         for placeable in &*this.scc.await? {
-            bools.push(is_placeable_self_async(*placeable));
+            if let Some(async_module) = &*placeable.get_async_module().await? {
+                if *async_module.is_self_async().await? {
+                    return Ok(Vc::cell(true));
+                }
+            }
         }
 
         for scc in &*this.scope.get_scc_children(this.scc).await? {
             // Because we generated SCCs there can be no loops in the children, so calling
             // recursively is fine.
-            bools.push(AsyncModuleScc::new(*scc, this.scope).is_async());
+            if *AsyncModuleScc::new(*scc, this.scope).is_async().await? {
+                return Ok(Vc::cell(true));
+            }
         }
 
-        Ok(Vc::<Bools>::cell(bools).any())
```

Comment:
Is this faster because we've implemented early exiting, or because we no longer need to build up this `Bools` vc just to await it? Does this have implications for the [just-approved](https://github.com/vercel/next.js/pull/55437) https://github.com/vercel/next.js/blob/35fd8bcb0388420618a6ffb08ad5e93f716016c1/packages/next-swc/crates/next-api/src/project.rs#L706

Reply:
early exist reduces the number of tasks. The Bools::any() also save a task, but it's only a single one.

Reply:
Completions is technically similar, but no early exit is possible here. We could avoid 1 task by not using Completions::completed(), but it's only 1 task and it has effects on the invalidation: Without Completions::completed() it would need to rerun the current function when any completion fires. With Completions::completed() it only reruns Completions::completed() and doesn't need to rerun the current function.

There is a difference in how often these methods are called `is_async` is called many many times (for every module), while `any_output_changed` is only called once.